chain/types/actor_event.go

```diff
@@ -0,0 +1,47 @@
+package types
+
+import (
+	"github.com/ipfs/go-cid"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/abi"
+)
+
+type ActorEventBlock struct {
+	// what value codec does client want to match on ?
+	Codec uint64 `json:"codec"`
+	// data associated with the "event key"
+	Value []byte `json:"value"`
+}
+
+type SubActorEventFilter struct {
+	Filter  ActorEventFilter `json:"filter"`
+	Prefill bool             `json:"prefill"`
+}
+
+type ActorEventFilter struct {
+	// Matches events from one of these actors, or any actor if empty.
+	// TODO: Should we also allow Eth addresses here?
+	// For now, this MUST be a Filecoin address.
+	Addresses []address.Address `json:"address"`
+
+	// Matches events with the specified key/values, or all events if empty.
+	// If the `Blocks` slice is empty, matches on the key only.
+	Fields map[string][]ActorEventBlock `json:"fields"`
+
+	// Epoch based filtering ?
+	// Start epoch for the filter; -1 means no minimum
+	MinEpoch abi.ChainEpoch `json:"minEpoch,omitempty"`
+
+	// End epoch for the filter; -1 means no maximum
+	MaxEpoch abi.ChainEpoch `json:"maxEpoch,omitempty"`
+}
```

Comment:
I think clients can just use 0 as `MinEpoch` to specify "earliest" and -1 in both `MinEpoch` and `MaxEpoch`s to signify no-ops.

Reply:
I'd vote for having one way to do a thing. How about:

* `<-1` is an error
* `-1` means "earliest you have"
* `>=0` means from precisely that epoch _and_ if you don't have events for that epoch then you should error

Or, is there a use-case for "this epoch, but later if you don't have that one"? Unfortunately that leads to lossy information for the user: you don't get to know whether the first epoch the remote gives you is the first it _has_ or the first that there _are_ events for.

Reply:
This makes sense to me.

Reply:
@rvagg Actually, replaced it with what we have for the existing ETH events API. Good to have this consistent across both the event APIs.